File: Level 2/Scoreboard Inference 2/javascript/Scoreboard_Inference_2.js
/**
 * @param {number} N
 * @param {number[]} S
 * @return {number}
 */
function getMinProblemCount(N, S) {
    let max = 0
    let secondMax = 0
    let hasOne = false
    let hasOneRemainder = false
    let hasTwoRemainder = false
    
    for (let i = 0; i < N; i++) {
        const score = S[i]
        
        if (score > max) {
            secondMax = max
            max = score
        }
        else if (score > secondMax) {
            secondMax = score
        }
        
        
        hasOne = hasOne || score == 1
        hasOneRemainder = hasOneRemainder || (score > 1 && (score % 3 == 1))
        hasTwoRemainder = hasTwoRemainder || (score % 3 == 2)
    }
    
    let ones = 0
    let twos = 0
    if (hasTwoRemainder) {
        if (hasOne) {
          ones = 1
          twos = 1
        }
        else if (hasOneRemainder) {
            if (secondMax == max - 1 && max % 3 == 1) {
                twos = 1
                ones = 1
            }
            else {
                twos = 2
            }
        }
        else {
            twos = 1
        }
    }
    else if (hasOneRemainder || hasOne) {
        ones = 1
    }
    
    let threes = Math.floor(max / 3)
    let remainingSum = twos * 2 + ones;
    if (max % 3 == 0 && remainingSum == 3) {
        threes--;
    }
    else if (max % 3 == 1 && remainingSum == 4) {
        threes--;
    }
    
    if (twos + ones > 2 || (max % 3 == 2 && twos == 0)) {
        throw Error("ow")
    }
    
    return threes + twos + ones;
}

File: Level 2/Scoreboard Inference 2/python/Scoreboard_Inference_2.py
from typing import List

def getMinProblemCount(N: int, S: List[int]) -> int:
    highestScore = 0
    secondHighestScore = 0
    containsScoreOf1 = False
    containsScoreWithOneRemainder = False
    containsScoreWithTwoRemainder = False
    
    for score in S:
        if score > highestScore:
            secondHighestScore = highestScore
            highestScore = score
        elif score > secondHighestScore and score < highestScore:
            secondHighestScore = score
        
        containsScoreOf1 = containsScoreOf1 or (score == 1)
        containsScoreWithOneRemainder = containsScoreWithOneRemainder or (score > 1 and (score % 3 == 1))
        containsScoreWithTwoRemainder = containsScoreWithTwoRemainder or (score % 3 == 2)

    ones = 0
    twos = 0
    if containsScoreWithTwoRemainder:
        # If there is a score with a remainder of 2, then it is always optimal to have a 2-point problem
        twos = 1
        if containsScoreOf1:
            # If there is a score of 1, then we must have a 1-point problem
            ones = 1
        elif containsScoreWithOneRemainder:
            if secondHighestScore == highestScore - 1 and secondHighestScore % 3 == 0:
                ones = 1
            else:
                twos += 1
    elif containsScoreWithOneRemainder or containsScoreOf1:
        # If no score has a remainder of 2 but there is a score with a remainder of 1, then it is always optimal to have a 1-point problem
        ones = 1

    threes = highestScore // 3  # Number of 3-point problems needed to reach the highest score
    remaining_sum = twos * 2 + ones  # Sum of points of problems that are not 3-point problems
    if highestScore % 3 == 0 and remaining_sum == 3:
        # If the highest score is a multiple of 3 and the sum of the remaining problems is 3, then we can remove one 3-point problem
        threes -= 1
    elif highestScore % 3 == 1 and remaining_sum == 4:
        # If the highest score has a remainder of 1 and the sum of the remaining problems is 4, then we can remove one 3-point problem
        threes -= 1

    # Return the total number of problems needed
    return threes + twos + ones